### PR TITLE
plugins/tiny-devicons-auto-colors: init

### PR DESCRIPTION
**Pull request by khaneliman**
Resolves https://github.com/nix-community/nixvim/issues/2381

~Attempted to see if it would work with mini mockDevicons and it still seemed to throw an error.~

Looked like a load order issue from dependencies, manually inserting setupCode in corresponding plugin's `luaConfig.post` until we can handle `mkOrder` or `dag` https://github.com/nix-community/nixvim/issues/2292 is available

### CODE REVIEW

**Comment by khaneliman** [comment]
I think it may be a load order thing. Gonna try something out.

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (MattSturgeon) [inline plugins/by-name/tiny-devicons-auto-colors/default.nix:105]: I'm imagining having a central "icon provider" module that's responsible for a single assertion and an option where providers can register themselves.

Maybe this would go in top-level `plugins/icons.nix`, or maybe `plugins/by-name/webdevicons/alternatives.nix` ?

Probably doesn't fit within this PR, I'm guessing best as a post-merge refactor?

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (MattSturgeon) [inline plugins/by-name/tiny-devicons-auto-colors/default.nix:113]: No need for `optional (cfg?settings)` here, that's used in the generic helper utils because there may not be a `settings` option declared. If an option exists it will _always_ be in `config`.

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (MattSturgeon) [inline plugins/by-name/tiny-devicons-auto-colors/default.nix:121]: Q: are you happy with this or were you planning to explore a `mkOrder` based solution?

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (khaneliman) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:113]: True, that makes sense.... I was just lazy and copy pasted this over. :P

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (khaneliman) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:105]: I do like that actually, I had thought about something similar recently with the introduction of different `cmp` plugins for the same reason.

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (khaneliman) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:121]: I can switch to using that I think, essentially we would set the priority for the icons providers to the same level and set this below that priority. Does that sound correct?

Comment (MattSturgeon) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:119]: Yeah, sounds correct. Obviously, higher number is lower priority.

My only concern it that by setting a priority for these three plugins, they'll _all_ end up in a different position from where they are currently.

Currently they're merged pseudorandomly within the "default" merge order prio.

Comment (khaneliman) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:119]: The only thing that gets a little messy, right now, is delegating the mini priority in the current form of the module. The `mini.icons` stuff gets bundled in the same luaConfig as the rest of their modules. Should we isolate that config and move it up or just move all the mini stuff to a higher order?

Comment (MattSturgeon) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:119]: My initial instinct is that it'd be nice to still keep the mini stuff together. I just hope it doesn't cause issues.

This will be no worse than we we migrated everything to `plugins/by-name`, effectively shuffling the merge order.

File: plugins/by-name/mini/default.nix
Comment (MattSturgeon) [inline plugins/by-name/mini/default.nix:9]: I'm minded to suggest we avoid using default prio numbers internally, that way end users can reliably go "before" us if we consistently use a prio after `(mkBefore null).priority`, for example

IIRC, before is 500, default is 1000, and after is 1500? So perhaps we should aim to be in the 7xx & 8xx range?

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (MattSturgeon) [inline plugins/by-name/tiny-devicons-auto-colors/default.nix:103]:
```suggestion
        message = ''
          Nixvim: Either `plugins.web-devicons` or `plugins.mini`* must be enabled to use `tiny-devicons-auto-colors`.
            *If using `plugins.mini`, you must enable the `icons` module and the `mockDevIcons` option.'';
```

**Review by MattSturgeon — APPROVED** [review]
I don't see any issues.

Left a suggestion for being less prescriptive in the assertion message. Do with it what you will :grin: 

Maybe let's give other maintainers chance to look over this before merging, especially regarding the re-ordered `configLocation`?

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (traxys) [reply plugins/by-name/tiny-devicons-auto-colors/default.nix:104]: I think we could have a similar thing for lazy loading

**Review by GaetanLepage — APPROVED** [review]
LGTM

File: plugins/by-name/tiny-devicons-auto-colors/default.nix
Comment (MattSturgeon) [inline plugins/by-name/tiny-devicons-auto-colors/default.nix:92]:
```suggestion
```

This comment is leftover from before the icon providers were re-ordered?

**Review by traxys — APPROVED** [review]

**Comment by khaneliman** [comment]
@mergify queue

**Comment by khaneliman** [comment]
@mergify dequeue

**Comment by khaneliman** [comment]
@mergify queue

**Pull request merged**
